### PR TITLE
Add a shutdownhook to remove jobs owned by the process

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #8590
by adding a shutdown hook that removes the last planned usage job on exit, if owned by the current process.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

#### How did you try to break this feature and the system with this change?

- start the usage server
- see the job record being created for the running process
- stop the server
- check that the record is no longer in the DB

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
If you remove the last record (which, in my head is some kind of protection mechanism, but I might be wrong) - don't we run into risk if the following happens:
- you start 2 usage servers at the same/similar time (systemctl start)
- both usage servers try to run the job (while only one should be the "owner of the job" and the other one will notice this and, currently, NOT execute the duplicate job - it will never execute the duplicate job as long as there is that last record which indicates that another/first usage server is the owner of the job)

?

**Comment by DaanHoogland** [comment]
> If you remove the last record (which, in my head is some kind of protection mechanism, but I might be wrong) - don't we run into risk if the following happens:
> 
>     * you start 2 usage servers at the same/similar time (systemctl start)
> 
>     * both usage servers try to run the job (while only one should be the "owner of the job" and the other one will notice this and, currently, NOT execute the duplicate job - it will never execute the duplicate job as long as there is that last record which indicates that another/first usage server is the owner of the job)
> 
> 
> ?

Yes, this is correct. The issue is when you stop the Usage server owning the job, it will no longer have an owner and no server will ever execute a job anymore. Therefore the owning process should remove their job on exit.

In the current situation, you have to delete the record manually and the risk is the same. There is just a manual action needed that you need to be aware off.

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 1]
```diff
@@ -2183,4 +2186,17 @@ protected void runInContext() {
             }
         }
     }
+    private class AbbandonJob extends Thread {
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:2189]:
```suggestion
    private class AbandonJob extends Thread {
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Review by vishesh92 — APPROVED** [review]
clgtm
didn't test

**Comment by vladimirpetrov** [comment]
Not sure if this can be avoided but when the usage server is killed (with -9, simulating crash), this machanism no longer works - the old entry stays in the db and a new one (with the new pid) is not created when the server is started again.

**Comment by DaanHoogland** [comment]
@vladimirpetrov I don't think it can but I'll have a google. does it lgty otherwise? We can create an improvement later.

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing

**Pull request merged**